x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_by_object_ids.ts

```diff
@@ -93,64 +67,78 @@ export const getExportByObjectIds = async (
-  const rules = await findRules({
-    rulesClient,
-    filter,
-    page: 1,
-    fields: undefined,
-    perPage: 10000,
-    sortField: undefined,
-    sortOrder: undefined,
-  });
+interface FetchRulesResult {
+  rules: ExportableRule[];
+  missingRuleIds: string[];
+}
 
-  const alertsAndErrors = objects.map(({ rule_id: ruleId }) => {
-    const matchingRule = rules.data.find((rule) => rule.params.ruleId === ruleId);
-    if (
-      matchingRule != null &&
-      hasValidRuleType(matchingRule) &&
-      matchingRule.params.immutable !== true
-    ) {
-      return {
-        statusCode: 200,
-        rule: transformRuleToExportableFormat(internalRuleToAPIResponse(matchingRule)),
-      };
-    } else {
-      return {
-        statusCode: 404,
-        missingRuleId: { rule_id: ruleId },
-      };
+const fetchRulesByIds = async (
+  rulesClient: RulesClient,
+  ruleIds: string[]
+): Promise<FetchRulesResult> =>
+  withSecuritySpan('fetchRulesByIds', async () => {
+    // It's important to avoid too many clauses in the request otherwise ES will fail to process the request
+    // with `too_many_clauses` error (see https://github.com/elastic/kibana/issues/170015). The clauses limit
+    // used to be set via `indices.query.bool.max_clause_count` but it's an option anymore. The limit is [calculated
+    // dynamically](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-settings.html) based
+    // on available CPU and memory but the minimum value is 1024.
+    // 1024 chunk size helps to solve the problem and use the maximum safe number of clauses.
+    const CHUNK_SIZE = 1024;
+    const processChunk = async (ids: string[]) => {
+      const rulesResult = await findRules({
+        rulesClient,
+        filter: `alert.attributes.params.ruleId: (${ids.join(' OR ')})`,
+        page: 1,
+        fields: undefined,
+        perPage: CHUNK_SIZE,
+        sortField: undefined,
+        sortOrder: undefined,
+      });
+      const rulesMap = new Map<string, PartialRule<RuleParams>>();
+
+      for (const rule of rulesResult.data) {
+        rulesMap.set(rule.params.ruleId, rule);
+      }
+
+      const rulesAndErrors = ids.map((ruleId) => {
+        const matchingRule = rulesMap.get(ruleId);
+
+        return matchingRule != null &&
+          hasValidRuleType(matchingRule) &&
+          matchingRule.params.immutable !== true
+          ? {
+              rule: transformRuleToExportableFormat(internalRuleToAPIResponse(matchingRule)),
+            }
+          : {
+              missingRuleId: ruleId,
+            };
+      });
+
+      return rulesAndErrors;
+    };
+
+    const ruleIdChunks = chunk(ruleIds, CHUNK_SIZE);
+    // We expect all rules to be processed here to avoid any situation when export of some rules failed silently.
+    // If some error happens it just bubbles up as is and processed in the upstream code.
+    const rulesAndErrorsChunks = await Promise.all(ruleIdChunks.map(processChunk));
```

Comment:
have you considered to use `p-map` with limited concurrency (2 or 3)?
In case if user tries to export 10,000, it would result in 9 simultaneous requests, each of which return 1,000 rules which can cause  strain on ES and Kibana.
Have you done any performance measurements of this use case?

When using `p-map`, I suspect most of cases wont have any performance impact with concurrency 2. Since, number of rules won't be great than 2k. But might be beneficial on large number of rules

Reply:
++ for that. We also have historically used `initPromisePool` for the same reasons/purpose, which also has a concurrency parameter. For example, it is used in the prebuilt rules API endpoint handlers.

Reply:
My initial implementation contained `p-map` but I changed it in favour of plain `Promise.all()`. After performing some tests locally w/ and w/o `p-map` I haven't noticed any problems. For sure I tested it locally and didn't create resource limited environment for ES. The rules I tested on were almost bare rules around 1KB in size. But let's calculate how much space  heavy 10K rules will take. Assume one rules takes 20KB and finally we get 20KB x 10K = 200MB which doesn't look like too much in isolation especially for a distributed database designed to work under high load.

I can imagine situation when a lot of people export simultaneously around 10K heavy rules and it can lead to much higher load while it could be stretched in time. In this case it's all delegated to ES to process requests on it's own. `Promise.all()` will give us fastest possible export limited only by ES. `p-map` with concurrency 2-3 will make some requests waiting and slow down UX.

@vitaliidm do you any real life ES performance metrics applicable which concern you? I don't mind to switch back to `p-map` if you think it's necessary here.

Reply:
> Assume one rules takes 20KB and finally we get 20KB x 10K = 200MB which doesn't look like too much in isolation especially for a distributed database designed to work under high load.

@maximpn I think the problem is not in any concrete value like 200MB, but in the fact that the current implementation will be creating spikes in: network traffic between ES and Kibana, ES memory consumption, Kibana server memory consumption, etc. And the size of the spikes will depend on: the number of existing rules x the number of users exporting rules more or less at the same time. So this doesn't feel like a good scalable solution.

Ideally, we need at least decouple from the number of existing rules and let the complexity/load depend only on the number of users exporting rules simultaneously. By using p-map or promise pool we'd trade speed for stability and scalability. Speed of export is not a concern or requirement, but from SDHs we know that our users have a lot of issues with performance and scalability.

Reply:
I finally replaced `Promise.all()` with `p-map`. It's hard to notice performance difference between concurrency 2 and 3 locally so I decided to stick to value 2.